### PR TITLE
Average of bin calculated

### PR DESCRIPTION
**Pull request by isaac-tfn**
The average of distribution bins is now calculated rather than just being the mid point of the bin.

Describe Changes
---

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `pyproject.toml`

### CODE REVIEW

**Comment by isaac-tfn** [comment]
@BenTaylor-TfN Updated for weighted average to be an extra column in distribution, and tests updated to pass

File: src/caf/toolkit/cost_utils.py [hunk 1]
```diff
@@ -78,6 +82,10 @@ def check_df_col_names(self) -> CostDistribution:
         for col_name, col_val in cols.items():
             if col_val not in self.df:
                 err_cols.update({col_name: col_val})
+        cols.update({"weighted_avg_col": getattr(self, "weighted_avg_col")})
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/cost_utils.py:85]: Why are you doing this here instead of on line 78 where the rest of the columns are defined?

File: src/caf/toolkit/cost_utils.py
Comment (BenTaylor-TfN) [inline src/caf/toolkit/cost_utils.py:1]: Do we also want to adjust / add / make a distinction between the property:
- avg_values (for the bins)
- weighted_avg_values (the new thing being added)

File: src/caf/toolkit/cost_utils.py [hunk 2]
```diff
@@ -252,6 +303,7 @@ def from_file(
         max_col: str = "max",
         avg_col: str = "ave",
         trips_col: str = "trips",
+        weighted_avg_col: str = "weighted_ave",
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/cost_utils.py:306]: Missing docstring for this. Do we also want to make this optional and default to the same as the average bin?

File: src/caf/toolkit/cost_utils.py [hunk 3]
```diff
@@ -141,6 +149,46 @@ def band_share_vals(self) -> np.ndarray:
...
+                "cost": pd.DataFrame(cost_matrix).stack(),
+                "demand": pd.DataFrame(matrix).stack(),
+            }
+        )
+        df["bin"] = pd.cut(df["cost"], bins=bin_edges)
+        df["weighted"] = df["cost"] * df["demand"]
+        grouped = df.groupby("bin", observed=False)[["weighted", "demand"]].sum().reset_index()
+        grouped["bin_centres"] = grouped["bin"].apply(lambda x: x.mid)
+        grouped["averages"] = grouped["weighted"] / grouped["demand"]
+        return grouped["averages"].fillna(grouped["bin_centres"].astype("float")).to_numpy()
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/cost_utils.py:190]: Need to come back and think about this in more detail. Might be one worth chatting about the best way to do this

File: src/caf/toolkit/cost_utils.py (hunk 1 above)
Comment (isaac-tfn) [reply src/caf/toolkit/cost_utils.py:85]: So it isn't included in the check for if the columns are in the df. This means that for input dists they won't error in the old format..

File: src/caf/toolkit/cost_utils.py [hunk 4]
```diff
@@ -78,6 +82,10 @@ def check_df_col_names(self) -> CostDistribution:
         for col_name, col_val in cols.items():
             if col_val not in self.df:
                 err_cols.update({col_name: col_val})
+        cols.update({"weighted_avg_col": getattr(self, "weighted_avg_col")})
+
+        if self.weighted_avg_col not in self.df.columns:
+            self.df[self.weighted_avg_col] = self.df[self.avg_col]
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/cost_utils.py:88]: Following on from previous comment

Very minor change. I've moved the whitespace down a line so the code reads a bit more like:
1. Check that the required columns are given
2. Add in the "weighted_avg_col" if it doesn't already exist.

```suggestion

        cols.update({"weighted_avg_col": getattr(self, "weighted_avg_col")})
        if self.weighted_avg_col not in self.df.columns:
            self.df[self.weighted_avg_col] = self.df[self.avg_col]
```

File: src/caf/toolkit/cost_utils.py [hunk 5]
```diff
@@ -79,6 +83,11 @@ def check_df_col_names(self) -> CostDistribution:
             if col_val not in self.df:
                 err_cols.update({col_name: col_val})
 
+        # Add in the weighted_avg_col if not already in df
+        cols.update({"weighted_avg_col": getattr(self, "weighted_avg_col")})
+        if self.weighted_avg_col not in self.df.columns:
+            self.df[self.weighted_avg_col] = self.df[self.avg_col]
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/cost_utils.py:89]: How easy would it be to add in a quick test to cover this case?

As this is expected default behaviour when the weighted col is missing, I'd like a test to flag if this ever changes in future

**Review by BenTaylor-TfN — APPROVED** [review]
